shell/browser/extensions/api/tabs/tabs_api.cc

```diff
@@ -130,4 +165,149 @@ bool TabsExecuteScriptFunction::ShouldInsertCSS() const {
   return false;
 }
 
+ExtensionFunction::ResponseAction TabsGetFunction::Run() {
+  std::unique_ptr<tabs::Get::Params> params(tabs::Get::Params::Create(*args_));
+  EXTENSION_FUNCTION_VALIDATE(params.get());
+  int tab_id = params->tab_id;
+
+  auto* contents = electron::api::WebContents::FromWeakMapID(
+      v8::Isolate::GetCurrent(), tab_id);
+  if (!contents)
+    return RespondNow(Error("No such tab"));
+
+  tabs::Tab tab;
+
+  tab.id.reset(new int(tab_id));
+  // TODO(nornagon): in Chrome, the tab URL is only available to extensions
+  // that have the "tabs" (or "activeTab") permission. We should do the same
+  // permission check here.
+  tab.url = std::make_unique<std::string>(
+      contents->web_contents()->GetLastCommittedURL().spec());
+
+  return RespondNow(ArgumentList(tabs::Get::Results::Create(std::move(tab))));
+}
+
+ExtensionFunction::ResponseAction TabsSetZoomFunction::Run() {
+  std::unique_ptr<tabs::SetZoom::Params> params(
+      tabs::SetZoom::Params::Create(*args_));
+  EXTENSION_FUNCTION_VALIDATE(params);
+
+  int tab_id = params->tab_id ? *params->tab_id : -1;
+  auto* contents = electron::api::WebContents::FromWeakMapID(
+      v8::Isolate::GetCurrent(), tab_id);
+  if (!contents)
+    return RespondNow(Error("No such tab"));
+
+  auto* web_contents = contents->web_contents();
+  GURL url(web_contents->GetVisibleURL());
+  std::string error;
+  if (extension()->permissions_data()->IsRestrictedUrl(url, &error))
+    return RespondNow(Error(error));
+
+  auto* zoom_controller = contents->GetZoomController();
+  double zoom_level =
+      params->zoom_factor > 0
+          ? blink::PageZoomFactorToZoomLevel(params->zoom_factor)
+          : blink::PageZoomFactorToZoomLevel(
+                zoom_controller->GetDefaultZoomFactor());
+
+  zoom_controller->SetZoomLevel(zoom_level);
+
+  return RespondNow(NoArguments());
+}
+
+ExtensionFunction::ResponseAction TabsGetZoomFunction::Run() {
+  std::unique_ptr<tabs::GetZoom::Params> params(
+      tabs::GetZoom::Params::Create(*args_));
+  EXTENSION_FUNCTION_VALIDATE(params);
+
+  int tab_id = params->tab_id ? *params->tab_id : -1;
```

Comment:
Can extract this common tab check used across all runners into a macro ?

Reply:
this is largely copied from chrome's version of this, and i think a macro's probably not the right way to refactor this code.

Reply:
Cool, yeah the macro idea was simply based on `EXTENSION_FUNCTION_VALIDATE` :)